accounting/bank/models.py

```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2017 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+Bank application models.
+"""
+
+from uuid import uuid4
+
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from djmoney.models.fields import CurrencyField
+from localflavor.au.models import AUBusinessNumberField
+from localflavor.generic.models import BICField, IBANField
+from vies.models import VATINField
+
+from accounting.account.models import Account, Address
+from accounting.bank.fields import AUBankStateBranchField
+
+
+class Bank(models.Model):
+    """
+    A bank to represent more details about bank accounts for users.
+    """
+
+    uuid = models.UUIDField(
+        blank=False, null=False, default=uuid4, editable=False, verbose_name=_("UUID"),
+        help_text=_("The universally unique identifier for this bank."))
+    name = models.CharField(
+        max_length=80,
+        help_text=_("The official name of the bank."))
+    address = models.ForeignKey(
+        Address, models.CASCADE, related_name='banks', blank=True, null=True,
+        help_text=_("The address of this bank."))
+
+    class Meta:
+        verbose_name = _('Bank')
+        verbose_name_plural = _('Banks')
+
+    def __str__(self):
+        """
+        Identify the bank by name and UUID.
+        """
+        return '{name} - {identifier}'.format(name=self.name, identifier=self.uuid)
+
+
+class BankAccount(models.Model):
+    """
+    A bank account stores details to help identify a particular user's bank account anywhere in the world.
+
+    Multiple bank accounts can belong to a single user account.
+
+    Because bank account identifiers differ from country to country (although we're moving towards standardizing
+    on IBAN & SWIFT/BIC), we let all bank account identifying details be optional, to let the user pick and choose
+    what makes sense for their case.
```

Comment:
Again, as mentioned on all@, we need to ask for the fields that Transferwise is going to require, and segregate cases which Transferwise don't handle. Otherwise people aren't going to enter what is needed exactly, and it's going to be a mess every time someone adds or updates their bank info.

Also we need to check that the bank account matches the country of the invoicing entity, so you probably need to create the transferwise recipient at this stage - and store the info there when we can use them, rather than in our DB.

Reply:
We need this information to generate the invoice with the details required on the invoice. It's not just about the TransferWise flow, but all flows, including invoice generation. We can't generate invoices without having at least some of these fields, which people say they require on their invoices.

Reply:
Why does that allows to more easily generate the invoices? They don't need the bank account details, this is only for transfers. Before duplicating info between two places (Transferwise and our DB), and thus having to handle a layer of cache, I would like to understand why this would be helpful. It's easier to just have one reference, Transferwise's API?

Reply:
So I don't have a particularly strong argument, just a bunch of little points:

1. Not getting completely tied to a 3rd party; this has a bunch of potential consequences of its own
1. Rate limits since we'd require burst calls; this led to issues even for Ocim with the GitHub API
1. Easier to get started for this first iteration; since I won't have time to make a TransferWise API client and work with it in the first iteration, this info will be manually entered, so it'll go in the DB

Reply:
> If we ever want to change provider we could retrieve the data at that point, through the same API. We would likely need to edit it anyway, since another provider would likely require slightly different information (international vs local accounts, different justification, etc.). I'm not too worried about this at this stage.

:+1: 

> When do we need to do burst calls? Can't the recipient id be used in place of the banking information in API calls to Transferwise?

I imagined we'd do burst calls to the TransferWise API for recipient details using the recipient ID every month. We need to do it, right, if we're not storing that info? And even if we store it, if we need to update it we'd do that on TransferWise, still requiring API calls. I have not tested their API yet, but if we have a big enough team, I wonder if they have burst limits that'll cause trouble. Maybe I'm misunderstanding something in your vision here.

> I don't think you need to implement the Transferwise API at this stage - a field could just store the transferwise recipient id, entered manually.

But then where would we get bank details from, if we don't implement the API client? I see now that you said we might not need bank details on the invoice, but I wonder if some people require this for their own invoices.

Reply:
Woops, @antoviaque ^^